### PR TITLE
Improve Raspberry Pi support.

### PR DESCRIPTION
**Pull request by abarisani**
This PR improves raspberry pi support by:

  * making the code more consistent with the rest of the codebase
  * making the pi specific packages more consistent among themselves
  * removing unused functions or duplicate constants
  * adding missing documentation
  * fixing broken linkx
  * adding required copyright and licensing notes

### CODE REVIEW

**Comment by abarisani** [comment]
@kenbell can you take a look?

**Comment by kenbell** [comment]
@abarisani - sure.  I didn't get a chance today.  I'll take a look over at the weekend.

File: soc/bcm2835/rng.go [hunk 1]
```diff
@@ -44,14 +50,18 @@ func getRandomData(b []byte) {
...
 
+	hw.status = PeripheralAddress(RNG_STATUS)
+	hw.data = PeripheralAddress(RNG_DATA)
+	hw.ctrl = PeripheralAddress(RNG_CTRL)
+
 	// Discard
-	reg.Write(PeripheralAddress(RNG_BASE+RNG_STATUS), warmupCount)
-	reg.Write(PeripheralAddress(RNG_BASE+RNG_CTRL), RNG_RBGEN)
+	reg.Write(hw.status, warmupCount)
+	reg.Set(hw.ctrl, CTRL_EN)
```
Comment (kenbell) [inline soc/bcm2835/rng.go:64]: `reg.Set` fails to correctly initialize the hardware - the register should be set to `0x1`, not have the first bit set it appears.  With reg.Set, tamago fails to initialize getting stuck in the loop waiting for random bytes to be available.
```suggestion
	reg.Write(hw.ctrl, CTRL_EN)
```

**Review by kenbell — CHANGES_REQUESTED** [review]
The RNG code needs a couple of fixes to allow tamago to startup.  I've added comments with suggested fixes.

I have added a couple of notes for stylistic / naming stuff.

There's also a latent bug in the code i contributed that could do with being fixed.

File: soc/bcm2835/timer.go [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
 // that can be found in the LICENSE file.
 
 package bcm2835
 
+import (
+	"time"
+)
+
+// WatchdogPeriod is the fixed 16us frequency of the BCM2835 watchdog.
+const WatchdogFreq = uint64(16 * time.Microsecond)
```
[the remark this answers is not in the document]
Comment (abarisani) [reply soc/bcm2835/timer.go]: Fixed in https://github.com/f-secure-foundry/tamago/pull/6/commits/86bf59abca3aab106a982c7041d59f7a1a17e175

File: soc/bcm2835/gpio.go [hunk 3]
```diff
@@ -85,67 +59,76 @@ func NewGPIO(num int) (*GPIO, error) {
...
-	gpio.SelectFunction(GPIOFunctionInput)
+	gpio.SelectFunction(GPIO_INPUT)
 }
 
-// SelectFunction selects the function of a GPIO line
-func (gpio *GPIO) SelectFunction(fn GPIOFunction) error {
-	if fn > GPIOFunctionAltFunction5 {
-		return fmt.Errorf("invalid GPIO function %d", fn)
+// SelectFunction selects the function of a GPIO line.
+func (gpio *GPIO) SelectFunction(n uint32) (err error) {
```
[the remark this answers is not in the document]
Comment (kenbell) [reply soc/bcm2835/gpio.go]: Thanks - that looks good.

**Comment by kenbell** [comment]
Thanks for updates.  Pi2 seems to be working fine.  PiZero I'm seeing corrupt output on the UART - I'll take a look.

**Comment by abarisani** [comment]
> Thanks for updates. Pi2 seems to be working fine. PiZero I'm seeing corrupt output on the UART - I'll take a look.

Ok, I'll wait for the update on the PiZero before merging then. Thanks!

**Review by kenbell — APPROVED** [review]
PiZero issue was due to a change on my end - i was missing setting in config.txt on my test system.  Needed to add `core_freq=250` to the Pi config.txt (https://www.raspberrypi.org/documentation/configuration/uart.md)

**Pull request merged**